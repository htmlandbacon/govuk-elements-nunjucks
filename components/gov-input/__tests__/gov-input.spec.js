'use strict';

const nunjucks = require('nunjucks');
const cheerio = require('cheerio');

describe('gov-input', () => {
  it('should set the ID as input-{name} if ID attr not supplied', () => {
    const name = 'full-name';
    const output = nunjucks.render('./components/gov-input/template.njk', {
      label: 'Full name',
      name
    });
    const $ = cheerio.load(output);
    const inputId = $('input').attr('id');

    expect(inputId).toBe(`input-${name}`);
    expect(output).toMatchSnapshot();
  });

  it('should use ID attribute value over generated input-{name} ID', () => {
    const id = 'my-id';
    const output = nunjucks.render('./components/gov-input/template.njk', {
      label: 'Full name',
      name: 'full-name',
      id
    });
    const $ = cheerio.load(output);
    const inputId = $('input').attr('id');

    expect(inputId).toBe(id);
    expect(output).toMatchSnapshot();
  });

  it('should add a form hint using the hint attribute', () => {
    const hint = 'All your names';
    const output = nunjucks.render('./components/gov-input/template.njk', {
      label: 'Full name',
      name: 'full-name',
      hint
    });
    const $ = cheerio.load(output);
    const formHintText = $('label span.form-hint').html();

    expect(formHintText).toBe(hint);
    expect(output).toMatchSnapshot();
  });

  it('should add maxlength to input using the maxlength attribute', () => {
    const maxlength = '20';
    const output = nunjucks.render('./components/gov-input/template.njk', {
      label: 'Full name',
      name: 'full-name',
      maxlength
    });
    const $ = cheerio.load(output);
    const inputMaxlength = $('input').attr('maxlength');

    expect(inputMaxlength).toBe(maxlength);
    expect(output).toMatchSnapshot();
  });

  it('should set the value of the input using the value attribute', () => {
    const value = 'Testy McTesterson';
    const output = nunjucks.render('./components/gov-input/template.njk', {
      label: 'Full name',
      name: 'full-name',
      value
    });
    const $ = cheerio.load(output);
    const inputValue = $('input').attr('value');

    expect(inputValue).toBe(value);
    expect(output).toMatchSnapshot();
  });

  it('should add error message and classes when passed an error object', () => {
    const error = 'Name is wrong';
    const name = 'full-name';
    const output = nunjucks.render('./components/gov-input/template.njk', {
      label: 'Full name',
      name,
      error
    });
    const $ = cheerio.load(output);
    const formGroupClasses = $('.form-group').attr('class');
    const errorMsg = $('label span.error-message').text();

    expect(formGroupClasses).toBe('form-group form-group-error');
    expect(errorMsg).toBe(error);
    expect(output).toMatchSnapshot();
  });
  describe('class options', () => {
    it('should add classes to input/label when supplied', () => {
      const classes = {label: 'test-label-class', input: 'test-input-class'};
      const output = nunjucks.render('./components/gov-input/template.njk', {
        label: 'Full name',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('.form-group label span').attr('class');
      const inputClasses = $('.form-group input').attr('class');

      expect(labelClasses).toBe('test-label-class');
      expect(inputClasses).toBe('form-control test-input-class');
      expect(output).toMatchSnapshot();
    });
    it('should default to class form-label-bold when label class is not supplied', () => {
      const classes = {input: 'test-input-class'};
      const output = nunjucks.render('./components/gov-input/template.njk', {
        label: 'Full name',
        classes
      });
      const $ = cheerio.load(output);
      const labelClasses = $('.form-group label span').attr('class');
      const inputClasses = $('.form-group input').attr('class');

      expect(labelClasses).toBe('form-label-bold');
      expect(inputClasses).toBe('form-control test-input-class');
      expect(output).toMatchSnapshot();
    });
  });
});

