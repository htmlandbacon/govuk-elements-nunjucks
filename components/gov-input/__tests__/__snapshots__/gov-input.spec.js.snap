// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gov-input class options should add classes to input/label when supplied 1`] = `
<div id="input-undefined-form" class="form-group">
  <label for="input-undefined">
    <span class="test-label-class">Full name</span>
  </label>
  <input class="form-control test-input-class" id="input-undefined" name=""
  type="text" value="" />
</div>
`;

exports[`gov-input class options should default to class form-label-bold when label class is not supplied 1`] = `
<div id="input-undefined-form" class="form-group">
  <label for="input-undefined">
    <span class="form-label-bold">Full name</span>
  </label>
  <input class="form-control test-input-class" id="input-undefined" name=""
  type="text" value="" />
</div>
`;

exports[`gov-input should add a form hint using the hint attribute 1`] = `
<div id="input-full-name-form" class="form-group">
  <label for="input-full-name">
    <span class="form-label-bold">Full name</span>
    <span class="form-hint">All your names</span>
  </label>
  <input class="form-control" id="input-full-name" name="full-name" type="text"
  value="" />
</div>
`;

exports[`gov-input should add error message and classes when passed an error object 1`] = `
<div id="input-full-name-form" class="form-group form-group-error">
  <label for="input-full-name">
    <span class="form-label-bold">Full name</span>
    <span class="error-message">Name is wrong</span>
  </label>
  <input class="form-control" id="input-full-name" name="full-name" type="text"
  value="" />
</div>
`;

exports[`gov-input should add maxlength to input using the maxlength attribute 1`] = `
<div id="input-full-name-form" class="form-group">
  <label for="input-full-name">
    <span class="form-label-bold">Full name</span>
  </label>
  <input class="form-control" id="input-full-name" maxlength="20" name="full-name"
  type="text" value="" />
</div>
`;

exports[`gov-input should set the ID as input-{name} if ID attr not supplied 1`] = `
<div id="input-full-name-form" class="form-group">
  <label for="input-full-name">
    <span class="form-label-bold">Full name</span>
  </label>
  <input class="form-control" id="input-full-name" name="full-name" type="text"
  value="" />
</div>
`;

exports[`gov-input should set the value of the input using the value attribute 1`] = `
<div id="input-full-name-form" class="form-group">
  <label for="input-full-name">
    <span class="form-label-bold">Full name</span>
  </label>
  <input class="form-control" id="input-full-name" name="full-name" type="text"
  value="Testy McTesterson" />
</div>
`;

exports[`gov-input should use ID attribute value over generated input-{name} ID 1`] = `
<div id="my-id-form" class="form-group">
  <label for="my-id">
    <span class="form-label-bold">Full name</span>
  </label>
  <input class="form-control" id="my-id" name="full-name" type="text" value=""
  />
</div>
`;
